Update `importdata` Script

mindlessroman opened the pull request:
Now when running the `importdata.sh` script, the script checks whether a provided database or collection already exists before attempting to create it or not. @code4clouds also made the suggested change of switching from sed to ed for speed gains when downloading our sample data.

Fixes #79.

manojvazirani commented on data/getdata.sh:
```diff
@@ -7,8 +7,8 @@ echo "Unzipping the data files (overwrites any existing .tsv files)..."
 echo
 echo "Removing IMDb '\N' values..."
-sed --in-place 's/\\N//g' *.tsv
+ed -s *.tsv <<< 's/\\N//g'
```
did you mean `sed` here?

code4clouds replied:
ed is another software for linux that can do replacement.  Its very fast to do these conversions compared with sed.  Its reduced minutes of waiting to seconds.

manojvazirani commented on data/getdata.sh:
```diff
@@ -7,8 +7,8 @@ echo "Unzipping the data files (overwrites any existing .tsv files)..."
 echo "IMDb data files ready:"
 ls *.tsv
-echo 
+echo
```
Just a minor comment that I see too many "echo " for new lines

mindlessroman replied:
I think having the new lines breaks up the wall of text that can happen. However, if folks find it distracting more than useful, I can remove them.

allydurks replied:
I prefer a `printf "\n\n\n"` with however many newlines you want, that way the echo echo echo echo etc doesn't look so distracting

manojvazirani commented on data/importdata.sh:
```diff
@@ -18,16 +18,40 @@ len=${#collections[@]}
+  if [[ "${dbExists}" = "false" ]]; then
+    echo "Creating database '${databaseName}'..."
+
+    az cosmosdb database create -g ${resourceGroup} -n ${cosmosName} --db-name ${databaseName} > /dev/null
```
Error handling needed.

manojvazirani commented on data/importdata.sh:
```diff
@@ -18,16 +18,40 @@ len=${#collections[@]}
+      echo "(${step} of ${len}) Creating collection '${collections[i]}'"
+
+      partition="/'\$v'/${keys[$i]}/'\$v'"
+      az cosmosdb collection create -g ${resourceGroup} -n ${cosmosName} --db-name ${databaseName} --collection-name ${collections[$i]} \
```
Error handling.

manojvazirani commented on data/importdata.sh:
```diff
@@ -41,26 +65,26 @@ import_data() {
 
-    mongoimport --host $hostName -u $user -p $password --ssl --sslAllowInvalidCertificates --type tsv --headerline \
-      --db $databaseName --collection ${collections[$i]} --numInsertionWorkers 40 --file ${files[$i]}
+    mongoimport --host ${hostName} -u ${user} -p ${password} --ssl --sslAllowInvalidCertificates --type tsv --headerline \
```
Possible error handling based on return code

andrewfrancisque commented on data/getdata.sh:
```diff
@@ -7,8 +7,8 @@ echo "Unzipping the data files (overwrites any existing .tsv files)..."
 echo "IMDb data files ready:"
 ls *.tsv
-echo 
+echo
```
```suggestion
echo
read -p "Press any key to continue... " -n1 -s
```

I think it's useful to have a wait statement, so that the user can read the messages on the screen, instead of having the window just disappear. Maybe one before as well, to make sure the user meant to have this script run.

I'm not sure if that's commonly done with shell scripts, but I think it would be useful for the user.

andrewfrancisque commented on data/getdata.sh:
```diff
@@ -7,8 +7,8 @@ echo "Unzipping the data files (overwrites any existing .tsv files)..."
 gunzip -vf *.gz
```
(GH isn't letting me comment on line 4, so doing here instead)

TIOLI: To Manoj's point about error handling, should we error check on the curl call? Literal question, because it's not likely, but possible, that the domain changes, the files are moved, the network is down, or the site is unavailable for some other reason.

andrewfrancisque commented on data/importdata.sh:
```diff
@@ -18,16 +18,40 @@ len=${#collections[@]}
+
+  dbExists="$(az cosmosdb database exists --db-name ${databaseName} -n ${cosmosName} --key ${password})"
+
+  if [[ "${dbExists}" = "false" ]]; then
```
TIOLI: Can we add a user prompt, to see if they wanted to create a new DB/Collection? I'm thinking about the case where I have a collection named `Example`, but I accidentally put `Exampl`.

manojvazirani replied:
+1 on above.

mindlessroman replied:
Definitely could do. If that makes sense in the flow - ideally they are running this after running the `load_env.sh` script where these are stored in an vars.env file. Which makes more sense to do though, doing a 'just to be sure...do you want to make a new one?' or doing a 'make some edits to the vars.env file and re-`source` them'....? I could be convinced by either.

code4clouds replied:
I like scripts to ask all the questions at the beginning to avoid having to monitor it after execution.  You know long-running scripts provide a window to get coffee/tea.

andrewfrancisque commented on data/load_env.sh:
```diff
@@ -58,113 +58,133 @@ azLogin() {
+	"""
+	For use when creating a new database
+
+	Displays a lsit of available database replication origins under the resource group
```
list*

andrewfrancisque replied:
Only requested change

mindlessroman commented:
Where I've run the load_env.sh previously
```shell
→ ./data/importdata.sh

Step 1. Confirming locations...
Is this the correct resource group name? [CadetDemo] Y,N (leave blank for Y):
Resource group confirmed.
Is this the correct CosmosDB account name? [demo-pjdb] Y,N (leave blank for Y):
Account name confirmed.

Step 2. Creating Cosmos DB database...

Checking whether the database 'IMDb' already exists...
Creating database 'IMDb'...
Database 'IMDb' has been created.

Step 3. Creating Cosmos DB collections...
Checking whether the collection 'titles' exists...
(1 of 3) Creating collection 'titles'
Collection 'titles' has been created.
Checking whether the collection 'names' exists...
(2 of 3) Creating collection 'names'
Collection 'names' has been created.
Checking whether the collection 'principals_mapping' exists...
(3 of 3) Creating collection 'principals_mapping'
Collection 'principals_mapping' has been created.
Step 4. Importing IMDb data to Cosmos DB...

(1 of 3) Importing data from file 'title.basics.tsv' to collection 'titles'...
2019-02-05T10:47:29.996-0500    connected to: demo-pjdb.documents.azure.com:10255
[... abridged ...]
2019-02-05T11:05:30.417-0500    imported 5611000 documents

Import to 'titles' is complete

(2 of 3) Importing data from file 'name.basics.tsv' to collection 'names'...
2019-02-05T11:05:30.740-0500    connected to: demo-pjdb.documents.azure.com:10255
[... abridged ...]
2019-02-05T11:29:49.615-0500    imported 9117045 documents

Import to 'names' is complete

(3 of 3) Importing data from file 'title.principals.tsv' to collection 'principals_mapping'...
2019-02-05T11:29:50.166-0500    connected to: demo-pjdb.documents.azure.com:10255
[... prematurely ended due to a network issue ...]
Step 5. Finished importing data. Cleaning up...
removed 'title.basics.tsv'
removed 'name.basics.tsv'
removed 'title.principals.tsv'

Step 6. Reducing throughput on Azure...

Setting 'titles' throughput to 1700 to reduce cost...
{
  "offer": {
    "_etag": "\"00004329-0000-0000-0000-5c59cea80000\"",
    "_rid": "8rZ1",
    "_self": "offers/8rZ1/",
    "_ts": 1549389480,
    "content": {
      "offerIsAutoScaleEnabled": false,
      "offerIsRUPerMinuteThroughputEnabled": false,
      "offerMinimumThroughputParameters": {
        "maxConsumedStorageEverInKB": 0,
        "maxCountOfSharedThroughputCollectionsEverCreated": 0,
        "maxThroughputEverProvisioned": 0
      },
      "offerThroughput": 1700,
      "userSpecifiedThroughput": 1700
    },
    "id": "8rZ1",
    "offerResourceId": "cpBLANnPiQM=",
    "offerType": "Invalid",
    "offerVersion": "V2",
    "resource": "dbs/cpBLAA==/colls/cpBLANnPiQM=/"
  }
}

Setting 'names' throughput to 1700 to reduce cost...
{
  "offer": {
    "_etag": "\"00004429-0000-0000-0000-5c59ceab0000\"",
    "_rid": "0No1",
    "_self": "offers/0No1/",
    "_ts": 1549389483,
    "content": {
      "offerIsAutoScaleEnabled": false,
      "offerIsRUPerMinuteThroughputEnabled": false,
      "offerMinimumThroughputParameters": {
        "maxConsumedStorageEverInKB": 0,
        "maxCountOfSharedThroughputCollectionsEverCreated": 0,
        "maxThroughputEverProvisioned": 0
      },
      "offerThroughput": 1700,
      "userSpecifiedThroughput": 1700
    },
    "id": "0No1",
    "offerResourceId": "cpBLAM8UXCM=",
    "offerType": "Invalid",
    "offerVersion": "V2",
    "resource": "dbs/cpBLAA==/colls/cpBLAM8UXCM=/"
  }
}

Setting 'principals_mapping' throughput to 1700 to reduce cost...
{
  "offer": {
    "_etag": "\"00004529-0000-0000-0000-5c59ceae0000\"",
    "_rid": "TdiW",
    "_self": "offers/TdiW/",
    "_ts": 1549389486,
    "content": {
      "offerIsAutoScaleEnabled": false,
      "offerIsRUPerMinuteThroughputEnabled": false,
      "offerMinimumThroughputParameters": {
        "maxConsumedStorageEverInKB": 0,
        "maxCountOfSharedThroughputCollectionsEverCreated": 0,
        "maxThroughputEverProvisioned": 0
      },
      "offerThroughput": 1700,
      "userSpecifiedThroughput": 1700
    },
    "id": "TdiW",
    "offerResourceId": "cpBLAJ8b+qw=",
    "offerType": "Invalid",
    "offerVersion": "V2",
    "resource": "dbs/cpBLAA==/colls/cpBLAJ8b+qw=/"
  }
}
This may take 10 minutes to be reflected in the portal.Complete!
```

manojvazirani commented on data/importdata.sh:
```diff
@@ -17,77 +18,185 @@ len=${#collections[@]}
+# import_data
+
+# printf "Step 5. Finished importing data. Cleaning up...\n"
+# delete_tsv_files
```
@mindlessroman: Did you miss this?

manojvazirani commented on data/load_env.sh:
```diff
@@ -58,113 +58,133 @@ azLogin() {
-		printf "Enter your subscription ID [%s]: " $currentSub
+
+	while ([[ -z "${subscriptionId}" ]]); do
+		printf "Enter your subscription ID [%s]: " "${currentSub}"
```
@mindlessroman ^^^ ?

mindlessroman replied:
I think this is resolved with a change from tabs to spaces?
